src/line.rs

```diff
@@ -1619,4 +1633,235 @@ mod tests {
         file.directory_index = 0;
         assert_eq!(file.directory(&header), None);
     }
+
+
+    fn new_registers() -> StateMachineRegisters {
+        let mut regs = StateMachineRegisters::default();
+        regs.reset(true);
+        regs
+    }
+
+    fn assert_exec_opcode(header: LineNumberProgramHeader<LittleEndian>,
+                          initial_registers: StateMachineRegisters,
+                          opcode: Opcode,
+                          expected_registers: StateMachineRegisters,
+                          expect_new_row: bool) {
+        let mut sm = StateMachine::new(header);
+        sm.registers = initial_registers;
+
+        let is_new_row = sm.execute(opcode);
+
+        assert_eq!(is_new_row, expect_new_row);
+        assert_eq!(sm.registers, expected_registers);
+    }
+
+    #[test]
+    fn test_exec_special_noop() {
+        let header = make_test_header(&[]);
+
+        let initial_registers = new_registers();
+        let opcode = Opcode::Special(16);
+        let expected_registers = initial_registers.clone();
+
+        assert_exec_opcode(header, initial_registers, opcode, expected_registers, true);
+    }
+
+    #[test]
+    fn test_exec_special_negative_line_advance() {
+        let header = make_test_header(&[]);
+
+        let mut initial_registers = new_registers();
+        initial_registers.line = 10;
+
+        let opcode = Opcode::Special(13);
```

Comment:
All of these special opcode tests are using the example matrix from the end of section 6.2.5.1.